Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift

```diff
@@ -682,24 +742,32 @@ extension NIOAsyncWriter {
         /* fileprivate */ internal mutating func writerDeinitialized() -> WriterDeinitializedAction {
             switch self._state {
             case .initial(_, let delegate):
-                // The writer deinited before writing anything.
-                // We can transition to finished and inform our delegate
-                self._state = .finished(sinkError: nil)
+                if self.finishOnDeinit {
+                    // The writer deinited before writing anything.
+                    // We can transition to finished and inform our delegate
+                    self._state = .finished(sinkError: nil)
 
-                return .callDidTerminate(delegate)
+                    return .callDidTerminate(delegate)
+                } else {
+                    preconditionFailure("Deinited NIOAsyncWriter without finishing")
+                }
 
             case .streaming(_, _, _, let suspendedYields, let delegate):
-                // The writer got deinited after we started streaming.
-                // This is normal and we need to transition to finished
-                // and call the delegate. However, we should not have
-                // any suspended yields because they MUST strongly retain
-                // the writer.
-                precondition(suspendedYields.isEmpty, "We have outstanding suspended yields")
-
-                // We can transition to finished directly
-                self._state = .finished(sinkError: nil)
+                if self.finishOnDeinit {
+                    // The writer got deinited after we started streaming.
+                    // This is normal and we need to transition to finished
+                    // and call the delegate. However, we should not have
+                    // any suspended yields because they MUST strongly retain
+                    // the writer.
+                    precondition(suspendedYields.isEmpty, "We have outstanding suspended yields")
+
+                    // We can transition to finished directly
+                    self._state = .finished(sinkError: nil)
 
-                return .callDidTerminate(delegate)
+                    return .callDidTerminate(delegate)
+                } else {
+                    preconditionFailure("Deinited NIOAsyncWriter without finishing")
```

Comment:
Users are going to see this message so I'd probably add some text like which exact method they're supposed to call when.